doc/dev/osd_internals/multi_object_write_transaction.rst

```diff
@@ -0,0 +1,117 @@
+==============================
+Multi-Object Write Transaction
+==============================
+
+Multi-object write transaction enables a group of write operations 
+on multi-objects complete atomically.
+
+Description of the Process
+--------------------------
+
+The high level process is,
+
+  1. Define the following class,
+
+     class *CEPH_RADOS_API MultiObjectWriteOperation*
+
+     {
+
+       std::map<std::string, ObjectWriteOperation*> slaves;
+
+       ObjectWriteOperation master;
+
+     };
+
+     based on this, client could send a group of ops on multi-objects 
+     through *MOSDOp* to the primary osd in charge of the *master* object 
+     which is explicitly specified by client, call this osd *MASTER*
+
+  #. *MASTER* receives the multi-object write operation request, performs 
+     dead lock check (detailed below), extracts *ObjectWriteOperations* 
+     by objects from *slaves*, and sends them to corresponding primary 
+     osds individually, calls those osds *SLAVE*, the transaction 
+     enters the *PREPARE* state
+
+  #. *SLAVE* receives *MOSDOp* from *MASTER*, note each op received by *SLAVE*
+     is a group of write operations on a single slave object. *SLAVE* 
+     performs dead lock check, does the normal *prepare_transaction()* stuff, 
+     if everything is ok, *SLAVE* will lock the object, and delay further 
+     read/write operations on this object until the transaction finishs. 
+     *SLAVE* sends a *QUERY* message to *MASTER*, and returns *MASTER* 
+     *EINPROGRESS*. The *QUERY* message is used for *MASTER* to notify *SLAVE* 
+     when to release transaction context, *MASTER* will not ack *QUERY* until 
+     itself has released transaction context
```

Comment:
This seems wrong.  master can't actually count on any of this because an interval change would cause the slave primary to forget the lock state, so the responses from the slaves are essentially useless.  Blocking reads and writes doesn't actually become reliable until you've persisted that a transaction is underway in a manner that survives an interval change (like performing a repop).  It seems like this initial prepare needs to be merged into the log step with a success response happening only once the transaction is logged durably.